// @archpulse/codepulse/dist/index.js
#!/usr/bin/env node
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const commander_1 = require("commander");
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const update_notifier_1 = __importDefault(require("update-notifier"));
const analyzer_1 = require("./analyzer");
const html_1 = require("./reporter/html");
const badge_1 = require("./commands/badge");
const output_1 = require("./commands/output");
const explain_1 = require("./explain");
const mcp_setup_1 = require("./mcp-setup");
const mcp_1 = require("./commands/mcp");
const scan_1 = require("./commands/scan");
const program = new commander_1.Command();
// Read package.json for version and name
const pkgPath = path.resolve(__dirname, '../package.json');
const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf8'));
const isWin = process.platform === 'win32';
// Check for updates
(0, update_notifier_1.default)({
    pkg,
    updateCheckInterval: 1000 * 60 * 60, // 1 hour
    distTag: 'latest'
}).notify({ isGlobal: true });
// Auto-configure MCP on first run
if ((0, mcp_setup_1.shouldRunMcpSetup)()) {
    const count = (0, mcp_setup_1.setupMcpConfigs)();
    if (count > 0) {
        const symbol = isWin ? 'v' : '✓';
        console.log(chalk_1.default.green(`\n  ${symbol} Automatically configured CodePulse as an MCP server for ${count} AI agent(s) on your PC!\n`));
    }
}
program
    .name('codepulse')
    .description('Deep code analysis for JS/TS/Python and more')
    .version(pkg.version);
program.configureHelp({
    subcommandTerm: (cmd) => chalk_1.default.cyan(cmd.name()),
    subcommandDescription: (cmd) => chalk_1.default.gray(cmd.description()),
    optionTerm: (option) => chalk_1.default.yellow(option.flags),
    optionDescription: (option) => chalk_1.default.gray(option.description),
    commandUsage: (command) => chalk_1.default.magenta(command.name() + ' ' + command.usage()),
    commandDescription: (command) => chalk_1.default.italic(command.description()),
});
program.addHelpText('before', `
${chalk_1.default.bold.blue('  ____           _      ____       _               ')}
${chalk_1.default.bold.blue(' / ___|___   __| | ___|  _ \\ _   _| |___  ___      ')}
${chalk_1.default.bold.blue('| |   / _ \\ / _` |/ _ \\ |_) | | | | / __|/ _ \\     ')}
${chalk_1.default.bold.blue('| |__| (_) | (_| |  __/  __/| |_| | \\__ \\  __/     ')}
${chalk_1.default.bold.blue(' \\____\\___/ \\__,_|\\___|_|    \\__,_|_|___/\\___|     ')}
`);
program.addHelpText('after', `
${chalk_1.default.bold('Examples:')}
  ${chalk_1.default.gray('$')} codepulse scan .
  ${chalk_1.default.gray('$')} codepulse stats src --json
  ${chalk_1.default.gray('$')} codepulse explain complexity
`);
program
    .command('mcp')
    .description('Start the Model Context Protocol (MCP) server for AI agents')
    .action(async () => {
    try {
        await (0, mcp_1.runMcpServer)();
    }
    catch (err) {
        console.error(chalk_1.default.red('MCP Server failed to start:'), err);
        process.exit(1);
    }
});
program
    .command('setup-mcp')
    .description('Manually trigger automatic MCP configuration for AI agents')
    .action(() => {
    const count = (0, mcp_setup_1.setupMcpConfigs)();
    if (count > 0) {
        console.log(chalk_1.default.green(`\n  ✓ Successfully configured CodePulse as an MCP server for ${count} AI agent(s)!\n`));
    }
    else {
        console.log(chalk_1.default.yellow('\n  ! No supported AI agent configurations found on this system.\n'));
    }
});
program
    .command('scan [dir]')
    .description('Analyze project and generate full HTML report')
    .option('--open', 'Open report in browser after generation')
    .option('--sarif', 'Generate SARIF report for CI/CD')
    .option('-d, --debug', 'Show detailed issues list')
    .option('--json', 'Output issues as JSON (CI-friendly)')
    .option('--focus <type>', 'Filter by issue type (dead-export|high-complexity|god-file|critical-node)')
    .option('--severity <level>', 'Filter by severity (info|warning|error)')
    .option('--max-issues <number>', 'Limit number of issues shown')
    .option('--fail-on <level>', 'Exit with code 1 if issues of this severity exist')
    .option('--group-by <field>', 'Group output by field (file|type|severity)')
    .option('--strict', 'Strict mode: treat warnings as errors, lower thresholds')
    .action(scan_1.runScan);
program
    .command('stats [dir]')
    .description('Print quick stats to console')
    .option('--json', 'Output as JSON')
    .action(async (dir = '.', opts) => {
    const absDir = path.resolve(dir);
    const spinner = (0, ora_1.default)('Analyzing...').start();
    try {
        const result = await (0, analyzer_1.analyze)(absDir);
        spinner.stop();
        if (opts.json) {
            console.log(JSON.stringify({
                totalFiles: result.totalFiles,
                totalLines: result.totalLines,
                avgComplexity: result.avgComplexity,
                issues: result.issues.length,
                errors: result.issues.filter(i => i.severity === 'error').length,
                warnings: result.issues.filter(i => i.severity === 'warning').length,
            }, null, 2));
        }
        else {
            (0, output_1.printStats)(result, absDir);
        }
    }
    catch (err) {
        spinner.fail('Failed');
        console.error(err);
    }
});
program
    .command('dead [dir]')
    .description('Show unused exports')
    .option('--json', 'Output as JSON')
    .action(async (dir = '.', opts) => {
    const absDir = path.resolve(dir);
    const spinner = (0, ora_1.default)('Detecting dead code...').start();
    try {
        const result = await (0, analyzer_1.analyze)(absDir);
        spinner.stop();
        if (opts.json) {
            const dead = result.issues.filter(i => i.type === 'dead-export');
            console.log(JSON.stringify(dead, null, 2));
        }
        else {
            (0, output_1.printDeadCode)(result);
        }
    }
    catch (err) {
        spinner.fail('Failed');
        console.error(err);
    }
});
program
    .command('graph [dir]')
    .description('Generate only the dependency graph SVG')
    .action(async (dir = '.') => {
    const absDir = path.resolve(dir);
    const spinner = (0, ora_1.default)('Building graph...').start();
    try {
        const result = await (0, analyzer_1.analyze)(absDir);
        const reportPath = (0, html_1.generateReport)(result, absDir);
        spinner.succeed(chalk_1.default.green('Graph saved!'));
        console.log(`  ${chalk_1.default.cyan(path.join(reportPath, 'graph.svg'))}\n`);
    }
    catch (err) {
        spinner.fail('Failed');
        console.error(err);
    }
});
program
    .command('badge [dir]')
    .description('Generate a quality badge SVG')
    .action(async (dir = '.') => {
    const absDir = path.resolve(dir);
    try {
        const result = await (0, analyzer_1.analyze)(absDir);
        const healthStats = {
            vulnerabilities: result.issues.filter(i => i.type === 'vulnerability').length,
            deadExports: result.deadExports.length,
            godFiles: result.godFiles.length,
            criticalFiles: result.criticalFiles.length,
            hotspots: result.hotspots,
            avgComplexity: result.avgComplexity
        };
        const score = (0, html_1.calculateHealthScore)(healthStats, result);
        const badgeSvg = (0, badge_1.generateBadge)(result, score);
        const badgePath = (0, badge_1.saveBadge)(badgeSvg, absDir);
        console.log(chalk_1.default.green(`\n  ✓ Badge generated: ${badgePath}`));
        console.log(chalk_1.default.gray(`    Score: ${score}/100\n`));
    }
    catch (err) {
        console.error(chalk_1.default.red('Failed to generate badge'), err);
    }
});
program
    .command('explain [topic]')
    .description('Explain what a detected issue means and how to fix it')
    .action((topic) => {
    const topics = Object.keys(explain_1.EXPLANATIONS);
    const isWin = process.platform === 'win32';
    const sym = {
        bullet: isWin ? '*' : '•',
        cross: isWin ? 'x' : '✗',
        check: isWin ? 'v' : '✓',
        line: isWin ? '-' : '─'
    };
    if (!topic) {
        console.log('\n' + chalk_1.default.bold.cyan('  CodePulse — Available Topics'));
        console.log(chalk_1.default.gray(sym.line.repeat(30)) + '\n');
        for (const key of topics) {
            const e = explain_1.EXPLANATIONS[key];
            console.log(`  ${chalk_1.default.cyan(key)}`);
            console.log(`    ${chalk_1.default.gray(e.short)}\n`);
        }
        console.log(chalk_1.default.gray(`  Usage: ${chalk_1.default.white('codepulse explain <topic>')}\n`));
        return;
    }
    const key = topic.toLowerCase();
    const entry = explain_1.EXPLANATIONS[key];
    if (!entry) {
        console.log(chalk_1.default.red(`\n  Unknown topic: "${topic}"`));
        console.log(chalk_1.default.gray(`  Available: ${topics.join(', ')}\n`));
        process.exit(1);
    }
    const { full } = entry;
    console.log('\n' + chalk_1.default.bold.cyan(sym.line.repeat(52)));
    console.log(chalk_1.default.bold.white(`  ${key.toUpperCase().replace(/-/g, ' ')}`));
    console.log(chalk_1.default.cyan(sym.line.repeat(52)) + '\n');
    console.log(`  ${chalk_1.default.white(full.description)}\n`);
    console.log(chalk_1.default.bold.yellow('  Detected when:'));
    for (const c of full.criteria) {
        console.log(`    ${chalk_1.default.gray(sym.bullet)} ${c}`);
    }
    console.log('\n' + chalk_1.default.bold.red('  Risks:'));
    for (const r of full.risks) {
        console.log(`    ${chalk_1.default.red(sym.cross)} ${r}`);
    }
    console.log('\n' + chalk_1.default.bold.green('  Recommended fixes:'));
    for (const f of full.fix) {
        console.log(`    ${chalk_1.default.green(sym.check)} ${f}`);
    }
    console.log('\n' + chalk_1.default.gray(sym.line.repeat(52)) + '\n');
});
program.parse(process.argv);
